utils/utf8.cc

```diff
@@ -99,20 +108,26 @@ static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t l
                          (byte1 >= 0xF1 && byte1 <= 0xF3) ||
                          // F4, 80..8F, 80..BF, 80..BF
                          (byte1 == 0xF4 && byte2 <= 0x8F))) {
-                    bytes = 4;
-                } else {
-                    return pos;
+                    return codepoint_status{.bytes_validated = codepoint_len};
                 }
-            } else {
-                return pos;
             }
-        } else {
-            return pos;
         }
+        return codepoint_status{.error = true};
+    }
+}
 
-        len -= bytes;
-        pos += bytes;
-        data += bytes;
+// 3x faster than boost utf_to_utf
+static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t len) {
+    size_t pos = 0;
+
+    while (len) {
+        auto cs = evaluate_codepoint(data, len);
+        pos += cs.bytes_validated;
+        data += cs.bytes_validated;
+        len -= cs.bytes_validated;
+        if (cs.error || cs.more_bytes_needed) {
+            return pos;
```

Comment:
I didn't understand what you're trying to do here or why.
It seems to me you are trying to return the end of the last full UTF-8 codepoint, perhaps so we can take the last few bytes plus a few bytes from the next buffer to create the last codepoint. But I think that's not needed! The only thing we need to check is that the next "more_bytes_needed" bytes in the next buffer are between 0x80 and 0xBF. That's it! We don't need to remember what the old buffer contained! We just need to remember more_bytes_needed and that's it.

Reply:
No. UTF8 forbids overlong sequences,  0x80-0xbf is too weak.

Reply:
I see. I didn't understand this earlier. I am still not sure how such "overlong" sequences can be caused by anything except leading zeros, i.e., 0x80, although I see now https://tools.ietf.org/html/rfc3629#section-10 has another example, and I guess this code was carefully written with that in mind, we we can't try to "improve" or "simplify" it as I have been suggesting to do :-(